### PR TITLE
ref(workflows): use a single job to run GCP tests

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

To be able to handle edge-cases and complex scenarios (launch step no failing, but test-result failint and not being able to re-run it), and for the sake of simplicity, this PR proposes using a single job for handling all the following scenarios:
- Test without cached states
- Test with Zebra cached state
- Test with Lightwalletd cached state

### Complex Code or Requirements

This is a very delicate change, even though is simplifying the process, as it has to consider multiple scenarios which were using multiple jobs and steps, and now it's using a single job and some steps have been merged.

## Solution

- Use variables like `$DISK_OPTION` and `$MOUNT_FLAGS` to allow flexibility when mounting the volumes.

## Review

- We should run most tests across this changes

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

Move the scripts out of the workflow.

### CODE REVIEW

**Review by teor2345 — DISMISSED** [review]
This needs a bug fix in the base branch.

**Comment by gustavovalverde** [comment]
This should be good to go!

**Comment by teor2345** [comment]
Would it help to check these workflows manually, to make sure they are still using the expected cached states? And writing the expected cached states?

What's the trickiest part of this change? (Or the most likely bug in it.)

File: .github/workflows/ci-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/ci-integration-tests-gcp.yml]: Optional: This change seems to just be re-ordering the jobs, can we remove it from this PR?
(This applies to the next change in the PR diff as well.)

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/sub-deploy-integration-tests-gcp.yml]: Nitpick: This change seems to be a mistake in the merge, the old docs were correct.

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/sub-deploy-integration-tests-gcp.yml]: Nitpick: Some tests use a Zebra and lightwalletd variable, so the plural is correct

```suggestion
      # These paths must match the variables used by the tests in Rust, which are also set in
```

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/sub-deploy-integration-tests-gcp.yml]: Nitpick: This comment accidentally got deleted rather than being moved.

```suggestion
          # Check if we need to mount for Lightwalletd state
          # lightwalletd-full-sync reads Zebra and writes lwd, so it is handled specially.
          if [[ "${{ inputs.needs_lwd_state }}" == "true" || "${{ inputs.test_id }}" == "lwd-full-sync" ]]; then
```

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/sub-deploy-integration-tests-gcp.yml:370]: 2 lines above this line we use `$DISK_NAME`, but it hasn't been written to `$GITHUB_ENV` by any previous step.

**Review by teor2345 — COMMENTED** [review]
This looks good! Thank you for this cleanup.

Is there a description of how the workflow works now? Can we add that to the top of this workflow?

It looks like you merged all 3 of the cases in the PR description into the same job, but also made some other changes to simplify pipe handling. Is there any other significant change here?

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (teor2345) [inline .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: This change might need a full sync test, I think the fastest test that uses this script is a testnet full sync.

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (gustavovalverde) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:370]: Yeah, this comes from the main branch. I'll create another PR to improve this. I don't think using $GITHUB_ENV will be the right approach, as this is executing in the the GCP VM context.

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (gustavovalverde) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: This behavior is already in the `main` branch: https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/deploy-gcp-tests.yml#L196-L197

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (gustavovalverde) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml]: Fixed with: [fix(ci): use better comment](https://github.com/ZcashFoundation/zebra/pull/7682/commits/a32b3742ebf75221d3a60d7a1c033c7fe3e0a789)

File: .github/workflows/sub-deploy-integration-tests-gcp.yml
Comment (gustavovalverde) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: By the way, I can't trigger a sync test unless we merge (as these are "new" workflows) https://github.com/ZcashFoundation/zebra/actions/workflows/ci-integration-tests-gcp.yml

Comment (teor2345) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: So do we want to delay this large CI change until after the release?
I'll leave it to you and @upbqdn to decide, because he is the release manager.

Comment (gustavovalverde) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: I'd prefer not to as this PR won't have an impact on the built image. But I don't mind if @upbqdn prefers not to include this PR; that's ok for me as we can merge it just after the release. We can have it approved and labeled to not merge. Whatever is less traumatic.

Comment (teor2345) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: My concern was more about unexpected CI failures delaying the release, but I also think we've done good testing here.

Comment (upbqdn) [reply .github/workflows/sub-deploy-integration-tests-gcp.yml:292]: This release will be big enough even without this change, so let's delay?

**Comment by teor2345** [comment]
I removed the `Run lwd-full-sync-test` branch protection rule because it very rarely runs in PRs, it runs in a scheduled job now.

**Review by teor2345 — APPROVED** [review]

**Comment by teor2345** [comment]
@gustavovalverde I just saw this $DISK_NAME fix follow-up again in my final review:
https://github.com/ZcashFoundation/zebra/pull/7682/files#r1357377327

It would be ok to copy the same script you use to find the disk name into that SSH command.

**Pull request merged**
